### ISSUE TITLE
SoftwareRequirement: environment modules changes to env not propagating to tool

### ISSUE THREAD

**Issue by rupertnash** [common-workflow-language/cwltool#1377]
## Expected Behavior

If I have a CWL file with a SoftwareRequirement that is resolved by an environment module, then I would expect that the environment variables set propagate to the tool execution.

## Actual Behavior
They do not: only `PATH` is modified by the environment module

## Workflow Code
Note this depends on having environment modules set up. Modify the cwl to load a module you have available:

env.cwl:
```
#!/usr/bin/env cwl-runner
# -*- mode: yaml; -*-

cwlVersion: v1.1
class: CommandLineTool
$namespaces:
  cwltool: http://commonwl.org/cwltool#

hints:
  SoftwareRequirement:
    packages:
      ncl:
        version: ["6.6.2"]

inputs: []

baseCommand: env

outputs:
  output:
    type: stdout
```
modules.yml:
```
- type: modules
  modulecmd: /usr/bin/modulecmd
```

cwltool runs OK:
```
 cwltool --beta-dependency-resolvers-configuration modules.yml env.cwl 
INFO /lustre/home/shared/dc118/sw/cwl/3.0.20201117141248/bin/cwltool 3.0.20201117141248
INFO Resolved 'env.cwl' to 'file:///lustre/home/dc118/nashvest/env.cwl'
INFO [job env.cwl] /tmp/7u0vetk_$ env > /tmp/7u0vetk_/d5514f0d054e63af38037aba76940ead71fd7324
Autoloading gcc/6.3.0
Autoloading gsl/2.6-gcc6
INFO [job env.cwl] completed success
{
    "output": {
        "location": "file:///lustre/home/dc118/nashvest/d5514f0d054e63af38037aba76940ead71fd7324",
        "basename": "d5514f0d054e63af38037aba76940ead71fd7324",
        "class": "File",
        "checksum": "sha1$488689425380b273e1e59a0dd3fdc884a57a842c",
        "size": 471,
        "path": "/lustre/home/dc118/nashvest/d5514f0d054e63af38037aba76940ead71fd7324"
    }
}
INFO Final process status is success
```

Module that should have been loaded:
```
$ module show ncl
-------------------------------------------------------------------
/lustre/sw/modulefiles/ncl/6.6.2:

conflict        ncl
Autoloading gcc/6.3.0
module load     gcc/6.3.0
Autoloading gsl/2.6-gcc6
module load     gsl/2.6-gcc6
setenv          NCARG_ROOT      /lustre/sw/ncl/ncl-6.6.2
prepend-path    PATH /lustre/sw/ncl/ncl-6.6.2/bin
prepend-path    LD_LIBRARY_PATH /lustre/sw/ncl/ncl-6.6.2/lib
prepend-path    LD_LIBRARY_PATH /lustre/sw/ncl/ncl-6.6.2/lib64
prepend-path    MANPATH /lustre/sw/ncl/ncl-6.6.2/man
-------------------------------------------------------------------
```

Actual environment 
```
$ cat d5514f0d054e63af38037aba76940ead71fd7324 
PATH=/lustre/sw/ncl/ncl-6.6.2/bin:/lustre/sw/gsl/2.6/bin:/lustre/sw/gcc/6.3.0/bin:/lustre/home/shared/dc118/sw/cwl/3.0.20201117141248/bin:/lustre/sw/spack-cirrus/opt/spack/linux-centos7-x86_64/gcc-8.2.0/git-2.21.0-rcchd4zgfdherdlklrr2y3amq7p73svi/bin:/lustre/sw/epcc-utils/bin:/opt/clmgr/sbin:/opt/clmgr/bin:/opt/sgi/sbin:/opt/sgi/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/c3/bin:/sbin:/bin
HOME=/tmp/7u0vetk_
TMPDIR=/tmp/9af437kp
```

If I change the command to say `cwltool --beta-dependency-resolvers-configuration modules.yml --preserve-environment NCARG_ROOT env.cwl ` then the output of `env` is unchanged 

## Your Environment
* cwltool version: 3.0.20201117141248

@mr-c and I have discussed briefly on gitter: https://gitter.im/common-workflow-language/common-workflow-language?at=5fbfb71303ddd74df2f90e59
Where Michael suggests pinging @jmchilton and @hmenager for comment?

**Comment by rupertnash** [comment]
Galaxy docs say:
>  A DependencyManager attempts to resolve named and versioned dependencies by
>  searching for them under a list of directories. Directories should be
>  of the form:
> 
>         $BASE/name/version/...
> 
>  and should each contain a file 'env.sh' which can be sourced to **make the
>  dependency available in the current shell environment**.

A brief review of the resolvers suggests that they all act in the bold text manner. This make me think that the way cwltool is trying to use this feature (getting the PATH from the modified shell environment and then taking total control of the environment) is a bit wrong

**Comment by rupertnash** [comment]
Maybe instead we should:
- construct an environment as cwltool does now
- run the shell commands produced by `DependencyManager`
- capture that environment
- use the captured environment to run the tool

Thoughts?

**Comment by rupertnash** [comment]
With the patch in the above PR, the output is now:
```
PATH=/lustre/sw/ncl/ncl-6.6.2/bin:/lustre/sw/gsl/2.6/bin:/lustre/sw/gcc/6.3.0/bin:/lustre/sw/spack-cirrus/opt/spack/linux-centos7-x86_64/gcc-8.2.0/git-2.21.0-rcchd4zgfdherdlklrr2y3amq7p73svi/bin:/lustre/sw/epcc-utils/bin:/opt/clmgr/sbin:/opt/clmgr/bin:/opt/sgi/sbin:/opt/sgi/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/c3/bin:/sbin:/bin
HOME=/tmp/_lt1ov3b
TMPDIR=/tmp/9_v8x4vd
LD_LIBRARY_PATH=/lustre/sw/ncl/ncl-6.6.2/lib64:/lustre/sw/ncl/ncl-6.6.2/lib:/lustre/sw/gsl/2.6/lib:/lustre/sw/gcc/6.3.0/lib64:/lustre/sw/gcc/6.3.0/lib
PATH_modshare=/lustre/sw/gsl/2.6/bin:1:/usr/bin:1:/usr/share/Modules/bin:1:/usr/local/bin:1:/opt/sgi/bin:1:/lustre/sw/ncl/ncl-6.6.2/bin:1:/opt/clmgr/bin:1:/lustre/sw/gcc/6.3.0/bin:1:/lustre/sw/epcc-utils/bin:1:/opt/sgi/sbin:1:/bin:1:/opt/clmgr/sbin:1:/sbin:1:/usr/sbin:1:/lustre/sw/spack-cirrus/opt/spack/linux-centos7-x86_64/gcc-8.2.0/git-2.21.0-rcchd4zgfdherdlklrr2y3amq7p73svi/bin:1:/usr/local/sbin:1:/opt/c3/bin:1
LOADEDMODULES_modshare=gcc/6.3.0:1:ncl/6.6.2:1:gsl/2.6-gcc6:1
CPATH_modshare=/lustre/sw/gsl/2.6/include:1
PKG_CONFIG_PATH_modshare=/lustre/sw/gsl/2.6/lib/pkgconfig:1
NCARG_ROOT=/lustre/sw/ncl/ncl-6.6.2
LD_LIBRARY_PATH_modshare=/lustre/sw/ncl/ncl-6.6.2/lib:1:/lustre/sw/gcc/6.3.0/lib:1:/lustre/sw/ncl/ncl-6.6.2/lib64:1:/lustre/sw/gsl/2.6/lib:1:/lustre/sw/gcc/6.3.0/lib64:1
LIBRARY_PATH_modshare=/lustre/sw/gsl/2.6/lib:1
PWD=/tmp/7j06k2fd
CMAKE_PREFIX_PATH=/lustre/sw/gsl/2.6
CPATH=/lustre/sw/gsl/2.6/include
_LMFILES__modshare=/lustre/sw/modulefiles/ncl/6.6.2:1:/lustre/sw/modulefiles/gsl/2.6-gcc6:1:/lustre/sw/modulefiles/gcc/6.3.0:1
ACLOCAL_PATH_modshare=/lustre/sw/gsl/2.6/share/aclocal:1
LIBRARY_PATH=/lustre/sw/gsl/2.6/lib
LOADEDMODULES=gcc/6.3.0:gsl/2.6-gcc6:ncl/6.6.2
CMAKE_PREFIX_PATH_modshare=/lustre/sw/gsl/2.6:1
MANPATH_modshare=/lustre/sw/gcc/6.3.0/share/man:1:/lustre/sw/ncl/ncl-6.6.2/man:1:/lustre/sw/gsl/2.6/share/man:1
SHLVL=1
ACLOCAL_PATH=/lustre/sw/gsl/2.6/share/aclocal
MANPATH=/lustre/sw/ncl/ncl-6.6.2/man:/lustre/sw/gsl/2.6/share/man:/lustre/sw/gcc/6.3.0/share/man
MODULEPATH=/lustre/home/dc118/shared/modules:/lustre/home/shared/dc118/modules:/lustre/sw/modulefiles:/usr/share/Modules/modulefiles:/etc/modulefiles:/usr/share/modulefiles
_LMFILES_=/lustre/sw/modulefiles/gcc/6.3.0:/lustre/sw/modulefiles/gsl/2.6-gcc6:/lustre/sw/modulefiles/ncl/6.6.2
PKG_CONFIG_PATH=/lustre/sw/gsl/2.6/lib/pkgconfig
_=/usr/bin/env
```

This may be sufficient, however the environment does gain more than the minimum required variables (e.g. all the `*_modshare` variables set by `modulecmd` to manage state as well as say `PWD`

**Comment by mr-c** [comment]
@rupertnash Great! I think you should filter out `PWD` and `_` in the script

**Comment by rupertnash** [comment]
Solved by #1378
